### ISSUE TITLE
Make Profiles Country Aware

### ISSUE THREAD

**Issue by MoKob** [Project-OSRM/osrm-backend#2772]
We have many issues that relate to specific defaults in different countries.
E.g. highways to allow pedestrians (India), `trunk`roads allowing bicycles (https://github.com/Project-OSRM/osrm-backend/issues/639), or the side of road we are driving on (https://github.com/Project-OSRM/osrm-backend/issues/2269).

Also default speed-values differ on a by country basis.
So allow a country-based set-up, we would need to pass country information along with every way and choose from a set of different default values.

Main issues to adjust in my view would be:
 - [ ] access whitelist ([e.g.](https://github.com/Project-OSRM/osrm-backend/blob/087080a6c3204669275280a68005e538e08ee4d9/profiles/car.lua#L9))
 - [ ] speed values ([e.g.](https://github.com/Project-OSRM/osrm-backend/blob/087080a6c3204669275280a68005e538e08ee4d9/profiles/car.lua#L19-L39)
 - [ ] table which side on the road is driven on based on country (adjust stuff like: [this](https://github.com/Project-OSRM/osrm-backend/blob/087080a6c3204669275280a68005e538e08ee4d9/profiles/car.lua#L148)
 - [ ] adjust [turn penalty](https://github.com/Project-OSRM/osrm-backend/blob/087080a6c3204669275280a68005e538e08ee4d9/profiles/car.lua#L525-L533) based on the road side

**Comment by MoKob** [comment]
https://github.com/Project-OSRM/osrm-backend/issues/1735 is another issue related to this. U-turns allowed in general, but not reasonable based on speeds.

**Comment by MoKob** [comment]
Access to certain tracks might also be based on countries? https://github.com/Project-OSRM/osrm-backend/issues/576

**Comment by TheMarex** [comment]
For a discussion about a possible profile level implementation:

https://github.com/Project-OSRM/osrm-backend/issues/333

**Comment by andrewharvey** [comment]
Would this include uturn restrictions. eg. in NSW uturns cannot be made a traffic lights unless a sign says you can [1]. Where no turn restriction exists in OSM should OSRM follow the region default restricting uturns?

[1] http://www.rms.nsw.gov.au/roads/safety-rules/road-rules/intersections.html

**Comment by daniel-j-h** [comment]
Yes @andrewharvey such a behavior would be desirable: but there's currently no code for parsing such signs for enabling uturns. To pull this off I think you need to

- set a large penalty for uturns in the profiles ([this branch would help not to screw with ETAs](https://github.com/Project-OSRM/osrm-backend/pull/2399))
- parse uturn signs in the profiles and pull them through the code base to [this](https://github.com/Project-OSRM/osrm-backend/blob/7cbb1807e7ff3c63527fdf5be8a09130a53f00c7/src/extractor/edge_based_graph_factory.cpp#L437-L444) location, enabling uturns when there is a sign by setting a low uturn penalty

This is unrelated to country-awareness, though. If you want to go that route and need more help, please open a separate ticket.

**Comment by andrewharvey** [comment]
@daniel-j-h 
> This is unrelated to country-awareness, though. If you want to go that route and need more help, please open a separate ticket.

In some countries you might be permitted to make uturns at traffic lights. This is more about what to make the default when nothing is tagged in OSM.

Sorry I can't help out with the development, just wanted to add this case as a feature request.

**Comment by daniel-j-h** [comment]
Related: access whitelist defaults on a per country basis ([ref.](https://github.com/Project-OSRM/osrm-backend/issues/576)):

See tables in: http://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Access-Restrictions

**Comment by TheMarex** [comment]
We pushed this back on our backlog, going to remove a milestone until we have some concrete idea when we can work on this.

**Comment by danpat** [comment]
Note that there's a 90% complete implementation in #4167 - I've been working on it when I can.
